ext/lists.go

```diff
@@ -277,31 +425,124 @@ func flatten(list traits.Lister, depth int64) ([]ref.Val, error) {
+func sortListByAssociatedKeys(list, keys traits.Lister) (ref.Val, error) {
 	listLength := list.Size().(types.Int)
+	keysLength := keys.Size().(types.Int)
+	if listLength != keysLength {
+		return nil, fmt.Errorf(
+			"@sortByAssociatedKeys() expected a list of the same size as the associated keys list, but got %d and %d elements respectively",
+			listLength,
+			keysLength,
+		)
+	}
 	if listLength == 0 {
 		return list, nil
 	}
-	elem := list.Get(types.IntZero)
+	elem := keys.Get(types.IntZero)
 	if _, ok := elem.(traits.Comparer); !ok {
 		return nil, fmt.Errorf("list elements must be comparable")
 	}
 
-	sorted := make([]ref.Val, 0, listLength)
+	sortedIndices := make([]ref.Val, 0, listLength)
 	for i := types.IntZero; i < listLength; i++ {
-		val := list.Get(i)
-		if val.Type() != elem.Type() {
+		if keys.Get(i).Type() != elem.Type() {
 			return nil, fmt.Errorf("list elements must have the same type")
 		}
-		sorted = append(sorted, val)
+		sortedIndices = append(sortedIndices, i)
 	}
 
-	sort.Slice(sorted, func(i, j int) bool {
-		return sorted[i].(traits.Comparer).Compare(sorted[j]) == types.IntNegOne
+	sort.Slice(sortedIndices, func(i, j int) bool {
+		iKey := keys.Get(sortedIndices[i])
+		jKey := keys.Get(sortedIndices[j])
+		return iKey.(traits.Comparer).Compare(jKey) == types.IntNegOne
 	})
 
+	sorted := make([]ref.Val, 0, listLength)
+
+	for _, sortedIdx := range sortedIndices {
+		sorted = append(sorted, list.Get(sortedIdx))
+	}
 	return types.DefaultTypeAdapter.NativeToValue(sorted), nil
 }
 
+// sortByMacro transforms an expression like:
+//
+//	mylistExpr.sortBy(e, -math.abs(e))
+//
+// into something equivalent to:
+//
+//	cel.bind(
+//	   __sortBy_input__,
+//	   myListExpr,
+//	   __sortBy_input__.@sortByAssociatedKeys(__sortBy_input__.map(e, -math.abs(e))
+//	)
+func sortByMacro(meh cel.MacroExprFactory, target ast.Expr, args []ast.Expr) (ast.Expr, *cel.Error) {
+	varIdent := meh.NewIdent("@__sortBy_input__")
+	varName := varIdent.AsIdent()
+
+	targetKind := target.Kind()
+	if targetKind != ast.ListKind &&
+		targetKind != ast.SelectKind &&
+		targetKind != ast.IdentKind &&
+		targetKind != ast.ComprehensionKind && targetKind != ast.CallKind {
+		return nil, meh.NewError(target.ID(), fmt.Sprintf("sortBy can only be applied to a list, identifier, comprehension, call or select expression"))
+	}
+
+	mapCompr, err := parser.MakeMap(meh, meh.Copy(varIdent), args)
+	if err != nil {
+		return nil, err
+	}
+	callExpr := meh.NewMemberCall("@sortByAssociatedKeys",
+		meh.Copy(varIdent),
+		mapCompr,
+	)
+
+	bindExpr := meh.NewComprehension(
+		meh.NewList(),
+		"#unused",
+		varName,
+		target,
+		meh.NewLiteral(types.False),
+		varIdent,
+		callExpr,
+	)
+
+	return bindExpr, nil
+}
+
+func distinctList(list traits.Lister) (ref.Val, error) {
+	listLength := list.Size().(types.Int)
+	if listLength == 0 {
+		return list, nil
+	}
+	uniqueList := make([]ref.Val, 0, listLength)
+	seen := make(map[ref.Val]bool, listLength)
```

Comment:
It won't be very fast, but I think it's more accurate to write the following:

```go
for i := types.IntZero; i < listLength; i++ {
    val := list.Get(i)
    seen := false
    for j := types.IntZero; j < types.Int(len(uniqueList)); j++ {
      if i = j {
         continue
      } 
      other := uniqueList[i]
      if val.Type() == other.Type() && val.Equal(other) == types.True {
        seen = true
        break
      }
    }
    if !seen {
      uniqueList = append(uniqueList, val)  
    }
}
```

Reply:
Done and added a test for the `[1, 1.0]` case.